patches/api/0421-Paper-Item-API.patch

```diff
@@ -0,0 +1,582 @@
++import io.papermc.paper.math.Position;
++import io.papermc.paper.property.MutablePropertyView;
++import io.papermc.paper.property.PropertyKey;
++import net.kyori.adventure.text.Component;
++import org.bukkit.Color;
++import org.bukkit.DyeColor;
++import org.bukkit.FireworkEffect;
++import org.bukkit.MusicInstrument;
++import org.bukkit.NamespacedKey;
++import org.bukkit.entity.Axolotl;
++import org.bukkit.inventory.ItemFlag;
++import org.bukkit.inventory.ItemStack;
++import org.bukkit.inventory.meta.BookMeta;
++import org.bukkit.inventory.meta.trim.ArmorTrim;
++import org.bukkit.potion.PotionData;
++import org.bukkit.potion.PotionEffect;
++import org.jetbrains.annotations.ApiStatus;
++
++import java.util.List;
++import java.util.Set;
++
++@ApiStatus.Experimental
++public interface ItemProperties {
++
++    // COMMON
++    PropertyKey<List<EnchantmentPair>> ENCHANTMENTS = PropertyKey.of("Enchantments", TypeFactory.parameterizedClass(List.class, EnchantmentPair.class));
++    PropertyKey<Boolean> UNBREAKABLE = PropertyKey.of("Unbreakable", Boolean.class);
++    PropertyKey<List<Namespaced>> CAN_DESTROY_BLOCK_LIST = PropertyKey.of("CanDestroy", TypeFactory.parameterizedClass(List.class, Namespaced.class)); // TODO: BETTER REPRESENTATION
++    PropertyKey<List<Namespaced>> CAN_PLACE_ON_BLOCK_LIST = PropertyKey.of("CanPlaceOn", TypeFactory.parameterizedClass(List.class, Namespaced.class)); // TODO: BETTER REPRESENTATION
++    PropertyKey<Set<ItemFlag>> HIDE_FLAGS = PropertyKey.of("HideFlags", TypeFactory.parameterizedClass(Set.class, ItemFlag.class));
++    PropertyKey<Integer> REPAIR_COST = PropertyKey.of("RepairCost", Integer.class);
++    PropertyKey<Integer> CUSTOM_MODEL_DATA = PropertyKey.of("CustomModelData", Integer.class);
++    PropertyKey<Integer> DAMAGE = PropertyKey.of("Damage", Integer.class);
++    PropertyKey<List<AttributePair>> ATTRIBUTES = PropertyKey.of("AttributeModifiers", TypeFactory.parameterizedClass(List.class, AttributePair.class));
++    PropertyKey<MutablePropertyView> BLOCK_STATE = PropertyKey.of("BlockStateTag", MutablePropertyView.class);
++    PropertyKey<MutablePropertyView> ENTITY_TAG = PropertyKey.of("EntityTag", MutablePropertyView.class);
++
++    // DISPLAY PROPERTIES
++    PropertyKey<MutablePropertyView> DISPLAY = PropertyKey.of("display", MutablePropertyView.class);
++
++    PropertyKey<Component> DISPLAY_NAME = PropertyKey.of("Name", Component.class);
++    PropertyKey<List<? extends Component>> LORE = PropertyKey.of("Lore", TypeFactory.parameterizedClass(List.class, Component.class));
++    PropertyKey<Color> COLOR = PropertyKey.of("color", Color.class);
++    PropertyKey<Color> MAP_COLOR = PropertyKey.of("MapColor", Color.class);
++
++    // Armor Meta
++    PropertyKey<ArmorTrim> ARMOR_TRIM = PropertyKey.of("Trim", ArmorTrim.class);
++
++    // TropicalFishBucketMeta
++    PropertyKey<TropicalFishVariant> TROPICAL_FISH_BUCKET_VARIANT = PropertyKey.of("BucketVariantTag", TropicalFishVariant.class);
```

Comment:
Technically, this is also missing properties for `NoAI`, `Silent`, `NoGravity`, `Glowing`, `Invulnerable`, and `Health`, but those could also be set via the entity tag. But then again, they are completely valid properties as far as the game is concerned.